honeybadgermpc/progs/jubjub.py

```diff
@@ -232,15 +232,13 @@ def __str__(self):
     terms = []
     p2i = p
     for i in range(len(bs)):
-        term = SharedPoint(context,
-                           p2i.x * bs[i],
-                           (p2i.y - 1) * bs[i] + p.curve.Field(1),
-                           p.curve)
-        terms.append(term)
+        x = p2i.x * bs[i]
+        y = (p2i.y - 1) * bs[i] + p.curve.Field(1)
+        terms.append(SharedPoint(context, x, y, p.curve))
         p2i = p2i.double()
 
     accum = terms[0]
-    for i in terms[1:]:
-        accum = await accum.add(i)
+    for term in terms[1:]:
```

Comment:
This will proceed sequentially, but we can do it in `log(n)` rounds. we can use the same trick we used in `Equality`--
```
while len(terms) > 1:
    left_terms, right_terms = terms[::2], terms[1::2]
    terms = await asyncio.gather(*[l.add(r) for (l, r) in zip(left_terms, right_terms)])

return terms[0]
```

Reply:
This may overlook the last element if  the number of elements is odd, so padding ZERO to `terms` at the start of each loop is to be added.

Reply:
Actually, there's no need to pad with 0-- if `left_terms` has more elements than `right_terms`, you just need to append `left_terms[-1]` to terms.

Reply:
oh right.